### PR TITLE
feat: Conf call in muted state default toast sqcall-186

### PR DESCRIPTION
**Pull request by johnranj**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQCALL-186" title="SQCALL-186" target="_blank">StorySQCALL-186</a>  [iOS] Inform the user they joined the call in muted state
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Provide a toast message to indicate the user has joins in muted state during conference call by default

### Solutions

Added a toast view with message as "When you join a conference call, you are initially muted" when the user attends the call in conference call in muted state

### Testing

To do

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Wire-iOS Tests/Calling/CallInfoRootViewControllerTests.swift [hunk 1]
```diff
@@ -278,4 +277,21 @@ final class CallInfoRootViewControllerTests: XCTestCase {
...
+    // MARK: - Toast Message
+    func testToastView() {
+        // given / when
+        sut = CallInfoRootViewController(configuration: defaultFixture.groupVideoEstablished(mockUsers: mockUsers), selfUser: mockSelfUser)
+        let toastConfig = ToastConfiguration(message: L10n.Localizable.Call.Toast.MutedOnJoin.message, colorScheme: ColorSchemeColor.utilityNeutral, variant: ColorSchemeVariant.light, dismissable: true, moreInfoAction: nil)
+        sut.presentToastView(config: toastConfig)
+        // then
+        verify(matching: sut)
+    }
+    func testToastViewWithNetworkQualityView() {
```
Comment (David-Henner) [inline Wire-iOS Tests/Calling/CallInfoRootViewControllerTests.swift:289]:
```suggestion
    }

    func testToastViewWithNetworkQualityView() {
```

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoRootViewController.swift [hunk 2]
```diff
@@ -140,3 +140,14 @@ final class CallInfoRootViewController: UIViewController, UINavigationController
...
 
 }
+// MARK: - Test Helpers
+
+extension CallInfoRootViewController {
+    // used by snapshot tests
+    func dismissToastView() {
+        Toast.hide()
+    }
+    func presentToastView(config: ToastConfiguration) {
```
Comment (David-Henner) [inline Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoRootViewController.swift:150]:
```suggestion
    }

    func presentToastView(config: ToastConfiguration) {
```

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift [hunk 3]
```diff
@@ -124,8 +124,21 @@ final class CallInfoViewController: UIViewController, CallActionsViewDelegate, C
         addChild(statusViewController)
         [statusViewController.view, accessoryViewController.view, actionsView].forEach(stackView.addArrangedSubview)
         statusViewController.didMove(toParent: self)
+        showMutedToastMessageIfNeeded()
     }
-
+    private func showMutedToastMessageIfNeeded() {
+        if case .established(let duration) = configuration.state,
+            duration <= 5.0,
+           configuration.isMuted, !self.isToastVisible {
```
Comment (David-Henner) [inline Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift:132]:
```suggestion
        if case .established(let duration) = configuration.state,
            duration <= 5.0,
            configuration.isMuted, 
            !isToastVisible {
```

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift [hunk 4]
```diff
@@ -124,8 +124,21 @@ final class CallInfoViewController: UIViewController, CallActionsViewDelegate, C
...
+            let toastConfig = ToastConfiguration(message: L10n.Localizable.Call.Toast.MutedOnJoin.message,
+                                                 colorScheme: ColorSchemeColor.utilityNeutral,
+                                                 variant: ColorSchemeVariant.light,
+                                                 dismissable: true, moreInfoAction: nil)
+            Toast.show(with: toastConfig)
+            self.isToastVisible = true
```
Comment (David-Henner) [inline Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift:138]:
```suggestion
            isToastVisible = true
```

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift [hunk 5]
```diff
@@ -124,8 +124,21 @@ final class CallInfoViewController: UIViewController, CallActionsViewDelegate, C
         addChild(statusViewController)
         [statusViewController.view, accessoryViewController.view, actionsView].forEach(stackView.addArrangedSubview)
         statusViewController.didMove(toParent: self)
+        showMutedToastMessageIfNeeded()
```
Comment (David-Henner) [inline Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift:127]: I think this still needs to be moved in `updateState()`, otherwise it doesn't make sense to use `isToastVisible`

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift [hunk 6]
```diff
@@ -69,7 +69,7 @@ final class CallInfoViewController: UIViewController, CallActionsViewDelegate, C
     private let statusViewController: CallStatusViewController
     private let accessoryViewController: CallAccessoryViewController
     private let actionsView = CallActionsView()
-
+    private var isToastVisible: Bool = false
```
Comment (David-Henner) [inline Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift:72]: This could be renamed for better clarity. We don't need to know if the toast is visible, we need to know if it has been shown. So I would suggest to rename it `hasMutedToastBeenShown`

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoRootViewController.swift (hunk 2 above)
Comment (David-Henner) [reply Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoRootViewController.swift:150]: we need a line between methods

File: Wire-iOS Tests/Calling/CallInfoRootViewControllerTests.swift (hunk 1 above)
Comment (David-Henner) [reply Wire-iOS Tests/Calling/CallInfoRootViewControllerTests.swift:289]: we need a line between methods

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift (hunk 4 above)
Comment (David-Henner) [reply Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift:138]: no need to reference `self`

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift [hunk 7]
```diff
@@ -124,8 +124,21 @@ final class CallInfoViewController: UIViewController, CallActionsViewDelegate, C
...
+           configuration.isMuted, !self.isToastVisible {
+            let toastConfig = ToastConfiguration(message: L10n.Localizable.Call.Toast.MutedOnJoin.message,
+                                                 colorScheme: ColorSchemeColor.utilityNeutral,
+                                                 variant: ColorSchemeVariant.light,
+                                                 dismissable: true, moreInfoAction: nil)
+            Toast.show(with: toastConfig)
+            self.isToastVisible = true
+        }
+    }
+
```
Comment (David-Henner) [inline Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift:141]:
```suggestion

```

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift [hunk 8]
```diff
@@ -125,7 +125,21 @@ final class CallInfoViewController: UIViewController, CallActionsViewDelegate, C
...
+    
+    private func showMutedToastMessageIfNeeded() {
+        if case .established(let duration) = configuration.state,
+            duration <= 5.0,
+            configuration.isMuted,
+            !hasMutedToastBeenShown {
+            let toastConfig = ToastConfiguration(message: L10n.Localizable.Call.Toast.MutedOnJoin.message,
+                                                 colorScheme: ColorSchemeColor.utilityNeutral,
+                                                 variant: ColorSchemeVariant.light,
+                                                 dismissable: true, moreInfoAction: nil)
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift:137]: You may be able to make this a little more readable by using type inference:

```suggestion
            let toastConfig = ToastConfiguration(message: L10n.Localizable.Call.Toast.MutedOnJoin.message,
                                                 colorScheme: .utilityNeutral,
                                                 variant: .light,
                                                 dismissable: true, 
                                                 moreInfoAction: nil)
```

**Review by johnxnguyen — COMMENTED** [review]
There are some swift lint violations that need fixing.

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift [hunk 9]
```diff
@@ -125,7 +125,21 @@ final class CallInfoViewController: UIViewController, CallActionsViewDelegate, C
         [statusViewController.view, accessoryViewController.view, actionsView].forEach(stackView.addArrangedSubview)
         statusViewController.didMove(toParent: self)
     }
-
+
```
Comment (David-Henner) [inline Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift:128]:
```suggestion

```

File: Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift [hunk 10]
```diff
@@ -125,7 +125,21 @@ final class CallInfoViewController: UIViewController, CallActionsViewDelegate, C
...
+            !hasMutedToastBeenShown {
+            let toastConfig = ToastConfiguration(message: L10n.Localizable.Call.Toast.MutedOnJoin.message,
+                                                 colorScheme: ColorSchemeColor.utilityNeutral,
+                                                 variant: ColorSchemeVariant.light,
+                                                 dismissable: true, moreInfoAction: nil)
+            Toast.show(with: toastConfig)
+            hasMutedToastBeenShown = true
+        }
+    }
+
```
Comment (David-Henner) [inline Wire-iOS/Sources/UserInterface/Calling/CallInfoViewController/CallInfoViewController.swift:142]:
```suggestion

```

**Review by David-Henner — APPROVED** [review]

**Review by johnxnguyen — APPROVED** [review]

**Comment by johnxnguyen** [comment]
@johnranj  FYI you have some test failures

File: Wire-iOS Tests/Calling/CallInfoRootViewControllerTests.swift [hunk 11]
```diff
@@ -276,22 +276,4 @@ final class CallInfoRootViewControllerTests: XCTestCase {
...
-        verify(matching: sut)
-    }
-    func testToastViewWithNetworkQualityView() {
-        // given / when
-        sut = CallInfoRootViewController(configuration: defaultFixture.groupVideoEstablishedPoorConnection, selfUser: mockSelfUser)
-        let toastConfig = ToastConfiguration(message: L10n.Localizable.Call.Toast.MutedOnJoin.message, colorScheme: ColorSchemeColor.utilityNeutral, variant: ColorSchemeVariant.light, dismissable: true, moreInfoAction: nil)
-        sut.presentToastView(config: toastConfig)
-        // then
-        verify(matching: sut)
-    }
```
Comment (johnxnguyen) [inline Wire-iOS Tests/Calling/CallInfoRootViewControllerTests.swift]: Why did you need to remove these tests?

Comment (johnranj) [reply Wire-iOS Tests/Calling/CallInfoRootViewControllerTests.swift]: I have written this test cases to test the toasts are rendering correctly on the window. But snapshottesting framework has a limitation to test it on the window, currently we are testing it on the views, we need to some workaround to handle this I assume

**Comment by johnxnguyen** [comment]
@johnranj once the checks are green, can we merge this?

**Comment by agisilaos** [comment]
@johnxnguyen I get the same errors on CI and I found the same errors in @sb88k PR's

**Comment by johnxnguyen** [comment]
The checks is complaining about a `wire-ios-request-strategy` version that couldn't be found. I'm looking into it.

**Comment by johnxnguyen** [comment]
@johnranj checks are green, you can merge now.
